File: Trabajo Practico 1/Hilos/C++/contadorDeCaracteres.cpp
#include <chrono>
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <thread>
#include <unistd.h> 
#include <vector>

using namespace std;

/*Compilacion y Ejecucion:
g++ -o contadorDeCaracteres ./contadorDeCaracteres.cpp
./contadorDeCaracteres rutaArchivo cantidadDeHilos*/

int total = 0;

void eliminarLineasVacias(const string& archivo_original, vector<string>& vector_lineas) 
{
    ifstream archivo_entrada(archivo_original);
    if (!archivo_entrada.is_open()) 
    {
        cerr << "Error al abrir el archivo de entrada." << endl;
        return;
    }
    string linea;
    while (getline(archivo_entrada, linea)) 
    {
        if (!linea.empty()) vector_lineas.push_back(linea);
    }
    archivo_entrada.close();
}

void contarCaracteresEnRango(vector<string>& vector_lineas, int inicio, int fin)
{
    sleep(10);
    int totalCaracteres = 0;
    if (inicio < 0) inicio = 0;
    if (fin >= vector_lineas.size()) fin = vector_lineas.size() - 1;
    for (int i = inicio; i <= fin; ++i)
    {
        totalCaracteres += vector_lineas[i].size(); 
    }
    cout << "Resultado Parcial: " + to_string(totalCaracteres) << endl;
    total += totalCaracteres;
    return;
}

void procesarArchivo(vector<string>& vector_lineas, int cantidad_de_hilos) 
{    
    int cantidad_lineas_por_archivo = vector_lineas.size() / cantidad_de_hilos;

    vector<thread> threads;

    for (int i = 0; i < cantidad_de_hilos;i++)
    {
        int indiceInferior = i*cantidad_lineas_por_archivo;
        int indiceSuperior = ((i+1)*cantidad_lineas_por_archivo) - 1;
        
        if(i == cantidad_de_hilos-1 && (vector_lineas.size() % cantidad_de_hilos != 0))
        {
            indiceSuperior += (vector_lineas.size() % cantidad_de_hilos);
        }
        threads.push_back(thread(contarCaracteresEnRango,ref(vector_lineas),indiceInferior,indiceSuperior));
    }

    for (auto& thread : threads)
    {
        thread.join();
    }
}

int main(int argc, char* argv[]) 
{
    cout << "PID: " << getpid() << endl;
    if (argc != 3) 
    {
        cerr << "Uso: " << argv[0] << " <archivo> <numero de threads>" << endl;
        return EXIT_FAILURE;
    }

    string nombre_archivo = argv[1];
    int cantidad_de_hilos = stoi(argv[2]);

    vector<string> vector_lineas;
    eliminarLineasVacias(nombre_archivo, vector_lineas);

    if(vector_lineas.size() < cantidad_de_hilos)
    {
        cout << "La cantidad de hilos supera la cantidad de lineas no vacias del archivo" << endl;
        return EXIT_FAILURE;
    } 

    auto start = chrono::high_resolution_clock::now();

    procesarArchivo(vector_lineas, cantidad_de_hilos);

    auto fin = chrono::high_resolution_clock::now();

    auto duracion = chrono::duration_cast<chrono::nanoseconds>(fin - start);

    cout << "Resultado Total: " + to_string(total) << endl;
    cout << "Tiempo de ejecución: " << (double)duracion.count()/1000000 << " ms" << endl;

    sleep(5);

    return EXIT_SUCCESS;
}

File: Trabajo Practico 1/Hilos/Python/contadorDeCaracteres.py
import argparse
import os
import time
import threading

VALIDACION_PARAMETROS_OK = 0
ERROR_PARAMETROS = -1
ERROR_ARCHIVO_INVALIDO = -2
ERROR_HILOS = -3

listaSubtotal = []

def validarParametros():
    parser = argparse.ArgumentParser(description="Contador de caracteres dado un archivo usando múltiples hilos.")
    parser.add_argument("rutaArchivo", type=str, help="Ruta del archivo.")
    parser.add_argument("cantidadDeHilos", type=int, help="Cantidad de hilos para procesar el archivo.")
    
    args = parser.parse_args()
    
    if not os.path.isfile(args.rutaArchivo):
        print(f"Error: El archivo '{args.rutaArchivo}' no existe o la ruta es inválida.")
        return None, None, ERROR_ARCHIVO_INVALIDO

    if args.cantidadDeHilos <= 0:
        print("Error: La cantidad de hilos debe ser un número positivo.")
        return None, None, ERROR_PARAMETROS
    
    return args.rutaArchivo, args.cantidadDeHilos, VALIDACION_PARAMETROS_OK

def eliminarLineasVacias(archivo, vectorLineas):
    with open(archivo, 'r') as f:
        for linea in f:
            if linea:
                vectorLineas.append(linea.strip()) # Elimina saltos de linea
                #vectorLineas.append(linea)

def contarCaracteresEnRango(vectorLineas, indiceSuperior, indiceInferior):
    subtotal = 0
    for i in range(indiceSuperior, indiceInferior + 1):
        subtotal += len(vectorLineas[i])
    listaSubtotal.append(subtotal)
    print(f"Resultado Parcial: {subtotal}")

def procesarArchivo(vectorLineas, cantidadDeHilos):
    cantidadDeLineasPorHilo = len(vectorLineas) // cantidadDeHilos
    threads = []
    
    for i in range(cantidadDeHilos):
        indiceInferior = i * cantidadDeLineasPorHilo
        indiceSuperior = ((i + 1) * cantidadDeLineasPorHilo) - 1
        if i == cantidadDeHilos - 1 and len(vectorLineas) % cantidadDeHilos:
            indiceSuperior += len(vectorLineas) % cantidadDeHilos
        thread = threading.Thread(target=contarCaracteresEnRango, args=(vectorLineas, indiceInferior, indiceSuperior))
        threads.append(thread)
        thread.start()
    
    for thread in threads:
        thread.join()

def main():
    pid = os.getpid()
    print(f"PID: {pid}")
    
    rutaArchivo, cantidadDeHilos, resultado = validarParametros()
    
    if resultado != VALIDACION_PARAMETROS_OK:
        return resultado
    
    vectorLineas = []
    eliminarLineasVacias(rutaArchivo, vectorLineas)
    
    if len(vectorLineas) < cantidadDeHilos:
        print("La cantidad de hilos supera la cantidad de lineas no vacias del archivo")
        return ERROR_HILOS
    
    inicio = time.time()
    
    procesarArchivo(vectorLineas, cantidadDeHilos)

    fin = time.time()

    duracion_milisegundos = (fin - inicio) * 1000 
    
    print(f"Tiempo de ejecución: {duracion_milisegundos:.2f} ms")

    print(f"Resultado Total: {sum(listaSubtotal)}")

if __name__ == "__main__":
    main()
